### PR TITLE
SAAS-12017: Add updateElementFolder

### PR DESCRIPTION
**Pull request by dorsev**
Add a new function `dumpElementsToFolder` which updates a “hybrid” folder - a folder containing a Salto workspace and an SFDX project.

It will get the changes to apply, it will call `dumpElementsToFolder` to update the SFDX part. And in the future, it should use the leftover changes to the Salto workspace

Additional context for the reviewer
No release notes

Release Notes:
None

User Notifications:
None

### CODE REVIEW

File: packages/core/src/core/adapter_format.ts [hunk 1]
```diff
@@ -243,3 +243,57 @@ export const syncWorkspaceToFolder = ({
...
+          errors: [
+            {
+              severity: 'Error' as const,
+              message,
+              detailedMessage: message,
+            },
+          ],
+        }
+      }
+      return dumpElementsToFolder({ baseDir, changes, elementsSource: adapterContext.elementsSource })
```
Comment (dorsev) [inline packages/core/src/core/adapter_format.ts:295]: should I implement now the ` and then apply the leftover changes to the salto workspace` ? or should this be done in a separate PR (together with sync method?)

File: packages/core/src/core/adapter_format.ts [hunk 2]
```diff
@@ -243,3 +243,57 @@ export const syncWorkspaceToFolder = ({
...
+        return {
+          errors: [
+            {
+              severity: 'Error' as const,
+              message,
+              detailedMessage: message,
+            },
+          ],
+        }
+      }
```
Comment (dorsev) [inline packages/core/src/core/adapter_format.ts:282]: DRY

File: packages/core/src/core/adapter_format.ts [hunk 3]
```diff
@@ -243,3 +243,57 @@ export const syncWorkspaceToFolder = ({
...
+export type UpdateElementFolderResult = {
+  errors: ReadonlyArray<SaltoError>
+}
+
+export const updateElementFolder = async ({
+  workspace,
+  baseDir,
+  changes,
+  accountName,
+}: UpdateElementFolderArgs): Promise<UpdateElementFolderResult> =>
```
Comment (dorsev) [inline packages/core/src/core/adapter_format.ts:263]: this currently does not return the `unAppliedChanges`. Is that okay? should I change this API currently?

File: packages/core/src/core/adapter_format.ts (hunk 2 above)
Comment (dorsev) [inline packages/core/src/core/adapter_format.ts:282]: I tried to "DRY" this but I could not think of a good enough API here.

File: packages/core/src/core/adapter_format.ts [hunk 4]
```diff
@@ -243,3 +243,57 @@ export const syncWorkspaceToFolder = ({
...
+        return {
+          errors: [
+            {
+              severity: 'Error' as const,
+              message,
+              detailedMessage: message,
+            },
+          ],
+        }
+      }
```
Comment (ori-moisis) [inline packages/core/src/core/adapter_format.ts:294]: 1. why is this an error?
2. if there is really a need to check for this use case, maybe do it before the whole initialization of the adapter?

File: packages/core/src/core/adapter_format.ts [hunk 5]
```diff
@@ -243,3 +243,57 @@ export const syncWorkspaceToFolder = ({
...
+      })
+      const { dumpElementsToFolder } = adapter
+      if (dumpElementsToFolder === undefined) {
+        const message = `Account ${accountName}'s adapter does not support writing a non-nacl format`
+        return {
+          errors: [
+            {
+              severity: 'Error' as const,
+              message,
+              detailedMessage: message,
```
Comment (ori-moisis) [inline packages/core/src/core/adapter_format.ts:278]: it doesn't make sense that the message and the detailed message are the same here (cc: @shayc331 
@dantal4  @kami-delaroz  - not sure if you did this in more places, this is not great, and if people continue to extend code as if these are always supposed to be the same, it will be harder to fix in the future...)

in my branch I changed this (in the other function) to
```typescript
const { loadElementsFromFolder, dumpElementsToFolder } = adapter
      if (loadElementsFromFolder === undefined) {
        return {
          errors: [
            {
              severity: 'Error' as const,
              message: 'Format not supported',
              detailedMessage: `Account ${accountName}'s adapter does not support loading a non-nacl format`,
            },
          ],
        }
      }
      if (dumpElementsToFolder === undefined) {
        return {
          errors: [
            {
              severity: 'Error' as const,
              message: 'Format not supported',
              detailedMessage: `Account ${accountName}'s adapter does not support writing a non-nacl format`,
            },
          ],
        }
      }
```

my branch is not ready yet, so could you please make this change here?

File: packages/core/test/core/adapter_format.test.ts [hunk 6]
```diff
@@ -353,3 +371,137 @@ describe('syncWorkspaceToFolder', () => {
...
+  const mockAdapterName = 'mock'
+  const type = new ObjectType({
+    elemID: new ElemID(mockAdapterName, 'type'),
+    fields: {
+      f: { refType: BuiltinTypes.STRING },
+    },
+  })
+
+  const instance = new InstanceElement('instance', type, { f: 'v' })
+  const changes = [toChange({ after: instance })]
```
Comment (ori-moisis) [inline packages/core/test/core/adapter_format.test.ts:385]: we should not create consts outside of tests / `beforeEach`, these create implicitly shared context between tests

File: packages/core/test/core/adapter_format.test.ts [hunk 7]
```diff
@@ -353,3 +371,137 @@ describe('syncWorkspaceToFolder', () => {
...
+  beforeEach(() => {
+    mockAdapter = createMockAdapter(mockAdapterName)
+    adapterCreators[mockAdapterName] = mockAdapter
+  })
+  afterEach(() => {
+    delete adapterCreators[mockAdapterName]
+  })
+  describe('when called with valid parameters', () => {
+    let workspace: Workspace
+    let accountName: string
```
Comment (ori-moisis) [inline packages/core/test/core/adapter_format.test.ts:397]: why is this needed in this scope?

File: packages/core/test/core/adapter_format.test.ts [hunk 8]
```diff
@@ -353,3 +371,137 @@ describe('syncWorkspaceToFolder', () => {
...
+        elementsSource: expect.anything(),
+      })
+    })
+  })
+
+  describe('when updating elements to folder returns an error', () => {
+    let result: SyncWorkspaceToFolderResult
+    let errors: SaltoError[]
+    let workspace: Workspace
+    const accountName = mockAdapterName
```
Comment (ori-moisis) [inline packages/core/test/core/adapter_format.test.ts:420]: inconsistent style (in the previous block this was an unneeded `let`, now it seems like an unneeded `const`)
(unneeded because it is only used in the `beforeEach`, no point in having it defined in a larger context)

File: packages/core/test/core/adapter_format.test.ts [hunk 9]
```diff
@@ -353,3 +371,137 @@ describe('syncWorkspaceToFolder', () => {
...
+          severity: 'Error',
+          message: 'something failed',
+          detailedMessage: 'something failed',
+        },
+      ]
+      workspace = mockWorkspace({
+        name: 'workspace',
+        elements: [],
+        accountToServiceName: { [accountName]: mockAdapterName },
+      })
```
Comment (ori-moisis) [inline packages/core/test/core/adapter_format.test.ts:433]: seems like this initialization of an empty workspace is repeated in multiple blocks
maybe do it once in an outer `beforeEach` block instead?

File: packages/core/src/core/adapter_format.ts (hunk 5 above)
Comment (shayc331) [reply packages/core/src/core/adapter_format.ts:278]: @ori-moisis it is an intermediate stage, currently the saas reads the `message` and I am changing it to read the `detailedMessage`. In the end, the UI will group errors by the `message` and it will be different than the `detailedMessage`

File: packages/core/src/core/adapter_format.ts (hunk 4 above)
Comment (dorsev) [reply packages/core/src/core/adapter_format.ts:294]: I wasn't sure of this. good point. removed

File: packages/core/test/core/adapter_format.test.ts (hunk 9 above)
Comment (dorsev) [reply packages/core/test/core/adapter_format.test.ts:433]: good point. changed

File: packages/core/test/core/adapter_format.test.ts (hunk 8 above)
Comment (dorsev) [reply packages/core/test/core/adapter_format.test.ts]: good point. changed

File: packages/core/test/core/adapter_format.test.ts (hunk 7 above)
Comment (dorsev) [reply packages/core/test/core/adapter_format.test.ts]: it's not. good point. changed

File: packages/core/test/core/adapter_format.test.ts (hunk 6 above)
Comment (dorsev) [reply packages/core/test/core/adapter_format.test.ts:385]: 👍 ack. changed. 
(moved to beforeEach)

**Comment by dorsev** [comment]
Thank you for the thorough review @ori-moisis 
I think I'm ready for round 2.

File: packages/core/src/core/adapter_format.ts (hunk 1 above)
Comment (dorsev) [reply packages/core/src/core/adapter_format.ts:280]: opened - SAAS-12078

File: packages/core/src/core/adapter_format.ts (hunk 3 above)
Comment (dorsev) [reply packages/core/src/core/adapter_format.ts:261]: yes 👍

File: packages/core/src/core/adapter_format.ts (hunk 5 above)
Comment (dorsev) [reply packages/core/src/core/adapter_format.ts]: good point @ori-moisis , I agree. added you suggested code there.

File: packages/core/src/core/adapter_format.ts [hunk 10]
```diff
@@ -243,3 +241,44 @@ export const syncWorkspaceToFolder = ({
...
+  baseDir: string
+  accountName: string
+  changes: ReadonlyArray<Change>
+}
+
+export type UpdateElementFolderResult = {
+  errors: ReadonlyArray<SaltoError>
+}
+
+export const updateElementFolder = async ({
```
Comment (ori-moisis) [inline packages/core/src/core/adapter_format.ts:256]: I missed this on the first round, but this doesn't need to be `async`
```suggestion
export const updateElementFolder = ({
```

File: packages/core/test/core/adapter_format.test.ts [hunk 11]
```diff
@@ -353,3 +372,94 @@ describe('syncWorkspaceToFolder', () => {
...
+      },
+    })
+
+    const instance = new InstanceElement('instance', type, { f: 'v' })
+    changes = [toChange({ after: instance })]
+    workspace = mockWorkspace({
+      name: 'workspace',
+      elements: [],
+      accountToServiceName: { [mockAdapterName]: mockAdapterName },
+    })
```
Comment (ori-moisis) [inline packages/core/test/core/adapter_format.test.ts:397]: note that you create an addition change of an instance, but the type of that instance is not defined in the workspace (i.e - it is not added, and does not exist in the workspace)
this is technically a possible scenario, but not a very valid one.
maybe it doesn't currently affect the results of this specific test, but I don't think you meant to test this...

moreover, in this test you are supposed to update a folder with changes from a workspace, but in this case the workspace does not contain the elements that were supposedly added to it.
this is actually not valid input here, right?

**Review by ori-moisis — APPROVED** [review]
a few nitpicks, other than that, code looks good

File: packages/core/src/core/adapter_format.ts (hunk 10 above)
Comment (dorsev) [reply packages/core/src/core/adapter_format.ts:256]: that's a good catch

File: packages/core/test/core/adapter_format.test.ts (hunk 11 above)
Comment (dorsev) [reply packages/core/test/core/adapter_format.test.ts]: good input. I've changed this. LMK if this makes more sense now? 🤔

Comment (ori-moisis) [reply packages/core/test/core/adapter_format.test.ts]: 👍
